src/ripple/app/consensus/RCLValidations.cpp

```diff
@@ -0,0 +1,337 @@
+#include <ripple/app/misc/ValidatorList.h>
+#include <ripple/basics/Log.h>
+#include <ripple/basics/StringUtilities.h>
+#include <ripple/basics/chrono.h>
+#include <ripple/consensus/LedgerTiming.h>
+#include <ripple/core/DatabaseCon.h>
+#include <ripple/core/JobQueue.h>
+#include <ripple/core/TimeKeeper.h>
+#include <memory>
+#include <mutex>
+#include <thread>
+
+namespace ripple {
+
+RCLValidations::RCLValidations(Application& app)
+    : Validations{ValidationParms{}, stopwatch()}
+    , app_(app)
+    , j_(app.journal("Validations"))
+{
+    staleValidations_.reserve(512);
+}
+
+NetClock::time_point
+RCLValidations::now() const
+{
+    return app_.timeKeeper().closeTime();
+}
+
+bool
+RCLValidations::add(STValidation::ref val, std::string const& source)
+{
+    auto const & signer = val->getSignerPublic();
+    auto const & hash = val->getLedgerHash();
+
+    // Ensure validation is marked as trusted if signer currently trusted
+    auto pubKey = app_.validators().getTrustedKey(signer);
+    if (!val->isTrusted() && pubKey)
+        val->setTrusted();
+
+    if (!val->isTrusted())
+    {
+        JLOG(j_.trace()) << "Node "
+                         << toBase58(TokenType::TOKEN_NODE_PUBLIC, signer)
+                         << " not in UNL st="
+                         << val->getSignTime().time_since_epoch().count()
+                         << ", hash=" << hash
+                         << ", shash=" << val->getSigningHash()
+                         << " src=" << source;
+    }
+
+    // If not currently trusted, see if signer is currently listed
+    if (!pubKey)
+        pubKey = app_.validators().getListedKey(signer);
+
+    bool currVal = isCurrent(val);
+    // only add trusted or listed
+    if (currVal && (val->isTrusted() || pubKey))
+    {
+        ScopedLockType sl(lock_);
+        auto res = Validations::add(now(), *pubKey, val);
+
+        // This is a duplicate validadtion
+        if (res == AddOutcome::repeat)
+            return false;
+        // This validation replaced a prior one with the same sequence number
+        if (res == AddOutcome::sameSeq)
+        {
+            auto const seq = val->getFieldU32(sfLedgerSequence);
+            JLOG(j_.warn())
+                << "Trusted node "
+                << toBase58(TokenType::TOKEN_NODE_PUBLIC, *pubKey)
+                << " published multiple validations for ledger " << seq;
+        }
+
+        currVal = res == AddOutcome::current;
+    }
+
+    JLOG(j_.debug()) << "Val for " << hash << " from "
+                     << toBase58(TokenType::TOKEN_NODE_PUBLIC, signer)
+                     << " added "
+                     << (val->isTrusted() ? "trusted/" : "UNtrusted/")
+                     << (currVal ? "current" : "stale");
+
+    if (val->isTrusted() && currVal)
+    {
+        app_.getLedgerMaster().checkAccept(
+            hash, val->getFieldU32(sfLedgerSequence));
+        return true;
+    }
+
+    // FIXME: This never forwards untrusted validations
```

Comment:
@scottschurr The idea was that we would have a certain number of validation slots with priority going to validators we trusted. Remaining slots might be allocated to validators that were listed by publishers we trusted but that we didn't choose to trust. The shorter term plan was just to forward untrusted validations if peers wanted them or if we had the ability/bandwidth to. None of that was implemented.